zerver/middleware.py

```diff
@@ -421,6 +422,13 @@ def process_request(self, request: HttpRequest) -> Optional[HttpResponse]:
             try:
                 request.realm = get_realm(subdomain)
             except Realm.DoesNotExist:
+                # Returning a 404 will get intercepted by the
+                # LocaleMiddleware and turned into a redirect to
+                # prepend /en/ -- unless the urlconf for the request
+                # has no locale bits in it:
```

Comment:
How exactly is that feature of LocaleMiddleware controlled?  We certainly return 404s in other places that don't result in such a redirect; I'm not sure why this one would be unique.

Reply:
Maybe the bug is that `invalid_realm.html` shouldn't have status code 404?

Reply:
Maybe this should be a 410 Gone response?  I'm not sure I've ever used one of those but the description seems to describe this situation:

```
10.4.11 410 Gone
The requested resource is no longer available at the server and no forwarding address is known. This condition is expected to be considered permanent. Clients with link editing capabilities SHOULD delete references to the Request-URI after user approval. If the server does not know, or has no facility to determine, whether or not the condition is permanent, the status code 404 (Not Found) SHOULD be used instead. This response is cacheable unless indicated otherwise.

The 410 response is primarily intended to assist the task of web maintenance by notifying the recipient that the resource is intentionally unavailable and that the server owners desire that remote links to that resource be removed. Such an event is common for limited-time, promotional services and for resources belonging to individuals no longer working at the server's site. It is not necessary to mark all permanently unavailable resources as "gone" or to keep the mark for any length of time -- that is left to the discretion of the server owner.

```

Reply:
[It's not configurable in any way, AFAICT](https://github.com/django/django/blob/d17b380653da5f95885ce53468fe7aac60672841/django/middleware/locale.py#L34) -- it just [activates if there are any LocalePrefixPatterns](https://github.com/django/django/blob/b9cf764be62e77b4777b3a75ec256f6209a57671/django/conf/urls/i18n.py#L32).  Those are generated by use of `i18n_patterns()`, [which we have](https://github.com/zulip/zulip/blob/d5b2381cfcd87f287079a04ec2679c236262357a/zproject/urls.py#L762).  It only pops up in URLs which are in the `i18n_urls` list, which can return a 404.  This turns out to be rare, but you can see the redirect happen on https://zulip.com/integrations/doc-html/missing which redirects to https://zulip.com/en/integrations/doc-html/missing.

Status 404 really does seem right.  I looked at all of the other status codes, and they all seem not-quite-right.  410 implies it _used_ to exist, does no longer, and will never change.  If a realm does not exist, that means it never has, but potentially might in the future -- we don't want browsers believing the lack of realm is permanent.  [421](https://httpstatuses.com/421) is not-right because that's very heavily tied into scheme and protocol mismatches.  And I don't think there are other valid codes.

Reply:
Oh, oops, I was misreading this as the "deactivated realm" code path -- agreed this should be 404.  

Looking at that code, though, it looks like there is a way to control it.  In particular, I think the issue is we should be setting `prefix_default_language = False` when we call `i18n_patterns()` in our `urls.py` file.  We don't intentionally use /en/ prefixes anywhere; looking at https://docs.djangoproject.com/en/3.1/topics/i18n/translation/#django.conf.urls.i18n.i18n_patterns, passing that argument is the documented way to disable this feature.

Reply:
We don't intentionally use `/en/`, but we do [document](https://zulip.readthedocs.io/en/latest/tutorials/life-of-a-request.html#views-serving-html-are-internationalized-by-server-path) that we have translated pages at e.g. https://zulip.com/zh-hans/features/.  If we pass `prefix_default_language`, that kills that feature, which may break existing links.

Reply:
Oh!  Nevermind, that's only for the default language -- _like the flag spells out, Alex_.  Will make that change.

Reply:
Unfortunately, this doesn't work. :(

Setting `prefix_default_language=False` means that a lack of language in the request path (e.g.  https://zulip.com/features/) [sets the request to be in the default language](https://github.com/django/django/blob/master/django/middleware/locale.py#L23), _even overriding things like `HTTP_ACCEPT_LANGUAGE`_.  This would effectively break all localization.

Reply:
And this behaviour looks to be quite intentional: https://github.com/django/django/pull/6264#issuecomment-194897558

Reply:
Yikes, I think Django made the wrong decision here in not providing a mechanism allow prefixing /LANG/ but never redirect to such prefixes.  We could consider reporting an issue and then doing a workaround by subclassing the locale middleware?  Or just merging the approach you have here with a comment linking to said issue, which is probably safer/simpler.

Reply:
Feel free to merge this tomorrow after opening a Django issue (if you agree we should) or adding a comment linking to this discussion for background on why we're doing this weird hack.

Reply:
[Filed upstream.](https://code.djangoproject.com/ticket/32005)

Can you give this another glance, particularly to see if the commit message explains the issue well enough?

Reply:
Looks great!